libp2p/io/trio.py

```diff
@@ -0,0 +1,33 @@
+import logging
+
+import trio
+from trio import SocketStream
+
+from libp2p.io.abc import ReadWriteCloser
+from libp2p.io.exceptions import IOException
+
+logger = logging.getLogger("libp2p.io.trio")
+
+
+class TrioReadWriteCloser(ReadWriteCloser):
+    stream: SocketStream
+
+    def __init__(self, stream: SocketStream) -> None:
+        self.stream = stream
+
+    async def write(self, data: bytes) -> None:
+        """Raise `RawConnError` if the underlying connection breaks."""
+        try:
+            await self.stream.send_all(data)
+        except (trio.ClosedResourceError, trio.BrokenResourceError) as error:
+            raise IOException(error)
```

Comment:
If you do this as `raise IOException from error` then you'll preserve the stacktrace information of the original exception which is typically beneficial.